### PR TITLE
Refactor some tests to use assertAll

### PR DESCRIPTION
**Pull request by franvila**
As I mention in a discussion in another thread, I suggest to avoid many assertions in one test or at least do it all together to avoid not testing something useful because the previous assertion fails.

To improve that, I decide to add `assertAll` to those tests that needed it and split negative tests to clarify the goal of each test and simplify them.

WDYT? @tombentley @SamBarker @k-wall

### CODE REVIEW

**Review by SamBarker — COMMENTED** [review]
I'm a big fan of "one assertion" per test, but I don't really see a lot of value in `assertAll`. I have always interpreted the one assertion to mean assert one behaviour, not as a single assertion statement. 

When writing and structuring tests I really like the BDD pattern of `Given, When, Then` for example (not using the actual API):

```java
public void clusterShouldHaveTwoBrokers() {
//Given
var clusterBuilder = new ClusterBuilder().withNumBrokers(2);

//When 
var cluster = clusterBuilder.build();

//Then 
assertThat(cluster.numBrokers)).is(2);
assertThat(cluster.boostrapServers).contains(",")
```

But that's not really the style of tests we have here. The tests here are much more stateful and I think they are trying to assert pre-conditions as well as actual behaviour.

**Comment by franvila** [comment]
> I'm a big fan of "one assertion" per test, but I don't really see a lot of value in `assertAll`. I have always interpreted the one assertion to mean assert one behaviour, not as a single assertion statement.
> 
> When writing and structuring tests I really like the BDD pattern of `Given, When, Then` for example (not using the actual API):
> 
> ```java
> public void clusterShouldHaveTwoBrokers() {
> //Given
> var clusterBuilder = new ClusterBuilder().withNumBrokers(2);
> 
> //When 
> var cluster = clusterBuilder.build();
> 
> //Then 
> assertThat(cluster.numBrokers)).is(2);
> assertThat(cluster.boostrapServers).contains(",")
> ```
> 
> But that's not really the style of tests we have here. The tests here are much more stateful and I think they are trying to assert pre-conditions as well as actual behaviour.

I agree with there are some pre-coditions and I could re-do this part, but using assertions for preconditions is not the best way to do it following the testing principles. Anyway, we could go with it.

I see the benefits of `assertAll` in the way that if the first assertion fails, we can continue testing the rest of assertions to be sure what is really failing in the test. If we don't use it, when the first assertions fails, the test stops there and nothing else is tested.

File: junit5-extension/src/test/java/io/kroxylicious/testing/kafka/junit5ext/ParameterExtensionTest.java [hunk 1]
```diff
@@ -45,44 +46,52 @@ public class ParameterExtensionTest extends AbstractExtensionTest {
     @Test
     public void clusterParameter(@BrokerCluster(numBrokers = 2) KafkaCluster cluster)
             throws ExecutionException, InterruptedException {
-        await().atMost(CLUSTER_FORMATION_TIMEOUT).untilAsserted(() -> assertEquals(2, describeCluster(cluster.getKafkaClientConfiguration()).nodes().get().size()));
         var dc = describeCluster(cluster.getKafkaClientConfiguration());
-        assertEquals(cluster.getClusterId(), dc.clusterId().get());
-        assertInstanceOf(InVMKafkaCluster.class, cluster);
+        assertAll(
+                () -> await().atMost(CLUSTER_FORMATION_TIMEOUT).untilAsserted(() -> assertEquals(2, describeCluster(cluster.getKafkaClientConfiguration()).nodes().get().size())),
```
[the remark this answers is not in the document]
Comment (franvila) [reply junit5-extension/src/test/java/io/kroxylicious/testing/kafka/junit5ext/ParameterExtensionTest.java]: done

File: junit5-extension/src/test/java/io/kroxylicious/testing/kafka/junit5ext/ParameterExtensionTest.java [hunk 2]
```diff
@@ -169,20 +203,21 @@ public void tlsClusterParameter(
...
-        assertTrue(matcher.find(),
-                "Expected '" + advertisedListener + "' to contain a match for " + listenerPattern.pattern());
+        var patternFound = matcher.find();
         var listenerName = matcher.group("listenerName");
         String protocolMap = brokerConfigs.get(KafkaConfig.ListenerSecurityProtocolMapProp()).value();
-        assertTrue(protocolMap.contains(listenerName + ":SSL"),
-                "Expected '" + protocolMap + "' to contain " + listenerName + ":SSL");
-
+        assertAll(
+                () -> assertFalse(bootstrapServer.contains(","), "expect a single bootstrap server"),
```
[the remark this answers is not in the document]
Comment (franvila) [reply junit5-extension/src/test/java/io/kroxylicious/testing/kafka/junit5ext/ParameterExtensionTest.java]: Agree

**Comment by tombentley** [comment]
This is the first time I've encountered `assertAll` (:scream:) and to be honest I don't immediately see the value of it. Sure I guess knowing that two assertions failed might be useful in some cases. But often enough (though not really in this case) I'm testing something stateful where is matters when an assertion holds in the lifecycle of an object. And I'm not a huge fan of all the extra `() ->` noise. 

All that said, I might just be very far behind the times in the best practices of writing tests (feel free to educate me if this is the case!). So if you both think this is valuable then I'm happy for it to be merged.

**Comment by SamBarker** [comment]
> I'm testing something stateful where is matters when an assertion holds in the lifecycle of an object

Hmm yeah about that... 

I really don't like tests that do that, test state after state. I think they encourage tests which try to verify too much in one go. The suite as a whole should verify all the behaviour. So to that end I think if there is merit to testing the "state transitions" then each one should be handled explicitly rather than in a test which goes through the full life-cycle. That largely stems from the one "assertion" per test practice. I use "" around assertion as that comes back to the point of this PR, which I'll come back to. I'm not completely against multi step tests, as they have their place, but I have a high bar for accepting them (i.e. they come in useful when dealing with code which wasn't built using TDD). 

I'm un-convinced by `assertAll`, I can see that it can have benefits in seeing all failures. Going back a few years the team I was on practised "one assertion per test" and we often wrote custom hamcrest matechers or assert  methods and I can see `assertAll` filling the same niche but I don't see the value in just dropping it in to wrap multiple unrelated assertions.

**Comment by franvila** [comment]
If no one is happy with this change I can close it, no problem. It was a suggestion only. Thanks for your feedback!
